File: apps/packages/ui/src/components/Option/MCPHub/mcpHubReadiness.ts
import type {
  McpHubExternalServer,
  McpHubToolRegistryEntry
} from "@/services/tldw/mcp-hub"

export type McpReadinessAction =
  | "add_server"
  | "edit_config"
  | "open_credentials"
  | "refresh_discovery"
  | "validate"
  | "view_details"
  | "open_tool_catalog"
  | "open_audit"

export type McpDisplayState =
  | "needs_setup"
  | "checking"
  | "ready"
  | "needs_attention"
  | "no_tools"
  | "stale"

export type McpReasonCode =
  | "not_configured"
  | "preflight_failed"
  | "discovery_not_run"
  | "auth_missing"
  | "runtime_unavailable"
  | "unreachable"
  | "discovery_failed"
  | "no_tools_returned"
  | "config_changed"
  | "catalog_expired"
  | "partial_capability"

export type McpCredentialState =
  | "not_required"
  | "required_missing"
  | "configured"
  | "legacy_fallback"
  | "unknown"

export type McpCurrentOperationHint =
  | string
  | {
      operation?: string
      label?: string
      startedAt?: string | null
      serverId?: string | null
    }

export type McpServerReadinessHint = {
  currentOperation?: McpCurrentOperationHint | null
  preflightFailed?: boolean
  discoverySucceededWithNoTools?: boolean
  unreachable?: boolean
  discoveryFailed?: boolean
  configChanged?: boolean
  catalogExpired?: boolean
  partialCapability?: boolean
}

export type McpServerReadiness = {
  serverId: string
  serverName: string
  displayState: McpDisplayState
  credentialState: McpCredentialState
  toolCount: number
  reasonCodes: McpReasonCode[]
  primaryReasonCode: McpReasonCode | undefined
  allowedActions: McpReadinessAction[]
  message: string
  currentOperation?: McpCurrentOperationHint
}

export type McpHubReadiness = {
  displayState: McpDisplayState
  reasonCodes: McpReasonCode[]
  primaryReasonCode: McpReasonCode | undefined
  allowedActions: McpReadinessAction[]
  message: string
  servers: McpServerReadiness[]
  totalServers: number
  readyServerCount: number
  checkingServerCount: number
  attentionServerCount: number
  noToolServerCount: number
  staleServerCount: number
}

type ReadinessHintsByServerId =
  | Record<string, McpServerReadinessHint | undefined>
  | Map<string, McpServerReadinessHint>

const REASON_PRIORITY: Record<McpReasonCode, number> = {
  not_configured: 0,
  auth_missing: 10,
  runtime_unavailable: 20,
  preflight_failed: 30,
  unreachable: 40,
  discovery_failed: 50,
  config_changed: 60,
  discovery_not_run: 70,
  no_tools_returned: 80,
  catalog_expired: 90,
  partial_capability: 100
}

const REASON_ACTIONS: Record<McpReasonCode, McpReadinessAction[]> = {
  not_configured: ["add_server"],
  preflight_failed: ["edit_config", "validate", "view_details"],
  discovery_not_run: ["refresh_discovery", "edit_config"],
  auth_missing: ["open_credentials", "view_details"],
  runtime_unavailable: ["edit_config", "view_details"],
  unreachable: ["edit_config", "refresh_discovery", "view_details"],
  discovery_failed: ["refresh_discovery", "view_details"],
  no_tools_returned: ["refresh_discovery", "view_details"],
  config_changed: ["refresh_discovery", "edit_config"],
  catalog_expired: ["refresh_discovery", "view_details"],
  partial_capability: ["open_tool_catalog", "view_details"]
}

const SERVER_READY_ACTIONS: McpReadinessAction[] = ["open_tool_catalog", "view_details"]
const HUB_READY_ACTIONS: McpReadinessAction[] = ["open_tool_catalog"]
const CHECKING_ACTIONS: McpReadinessAction[] = ["view_details"]

const sortReasonCodes = (reasonCodes: McpReasonCode[]): McpReasonCode[] =>
  [...new Set(reasonCodes)].sort(
    (left, right) =>
      REASON_PRIORITY[left] - REASON_PRIORITY[right] || left.localeCompare(right)
  )

const uniqueActions = (actions: McpReadinessAction[]): McpReadinessAction[] => {
  const seenActions = new Set<McpReadinessAction>()

  return actions.filter((action) => {
    if (seenActions.has(action)) {
      return false
    }

    seenActions.add(action)
    return true
  })
}

const getActionsForReasons = (
  reasonCodes: McpReasonCode[],
  fallbackActions: McpReadinessAction[] = []
): McpReadinessAction[] => {
  if (reasonCodes.length === 0) {
    return uniqueActions(fallbackActions)
  }

  return uniqueActions(reasonCodes.flatMap((reasonCode) => REASON_ACTIONS[reasonCode]))
}

const getDisplayStateForReason = (
  primaryReasonCode: McpReasonCode | undefined
): McpDisplayState => {
  switch (primaryReasonCode) {
    case undefined:
    case "partial_capability":
      return "ready"
    case "not_configured":
      return "needs_setup"
    case "config_changed":
    case "catalog_expired":
      return "stale"
    case "no_tools_returned":
      return "no_tools"
    case "discovery_not_run":
    case "preflight_failed":
    case "auth_missing":
    case "runtime_unavailable":
    case "unreachable":
    case "discovery_failed":
      return "needs_attention"
  }
}

const getOperationLabel = (currentOperation: McpCurrentOperationHint): string => {
  if (typeof currentOperation === "string") {
    return currentOperation
  }

  return currentOperation.label ?? currentOperation.operation ?? "Operation"
}

const getCredentialMessage = (credentialState: McpCredentialState): string => {
  switch (credentialState) {
    case "not_required":
      return " No credentials required."
    case "required_missing":
      return " Required credentials are missing."
    case "configured":
      return " Credentials are configured."
    case "legacy_fallback":
      return " Legacy server-level secret fallback is configured."
    case "unknown":
      return ""
  }
}

const getMessageForServer = ({
  displayState,
  primaryReasonCode,
  credentialState,
  toolCount,
  currentOperation
}: {
  displayState: McpDisplayState
  primaryReasonCode: McpReasonCode | undefined
  credentialState: McpCredentialState
  toolCount: number
  currentOperation?: McpCurrentOperationHint
}): string => {
  if (displayState === "checking" && currentOperation) {
    return `${getOperationLabel(currentOperation)} is in progress.${getCredentialMessage(
      credentialState
    )}`
  }

  const credentialMessage = getCredentialMessage(credentialState)

  switch (primaryReasonCode) {
    case undefined:
      return `Server is ready with ${toolCount} available tool${toolCount === 1 ? "" : "s"}.${credentialMessage}`
    case "not_configured":
      return "Add an MCP server to begin setup."
    case "preflight_failed":
      return `Preflight validation failed.${credentialMessage}`
    case "auth_missing":
      return `Credentials are required before this server can be used.${credentialMessage}`
    case "runtime_unavailable":
      return `The configured runtime is not available.${credentialMessage}`
    case "unreachable":
      return `The server is unreachable.${credentialMessage}`
    case "discovery_failed":
      return `Tool discovery failed.${credentialMessage}`
    case "config_changed":
      return `Configuration changed after the last discovery run.${credentialMessage}`
    case "catalog_expired":
      return `The tool catalog is expired.${credentialMessage}`
    case "discovery_not_run":
      return `Run discovery to populate this server's tool catalog.${credentialMessage}`
    case "no_tools_returned":
      return `Discovery completed, but the server returned no tools.${credentialMessage}`
    case "partial_capability":
      return `Server is ready, but some capabilities need review.${credentialMessage}`
  }
}

const getMessageForHub = ({
  displayState,
  primaryReasonCode,
  totalServers,
  readyServerCount,
  currentOperation
}: {
  displayState: McpDisplayState
  primaryReasonCode: McpReasonCode | undefined
  totalServers: number
  readyServerCount: number
  currentOperation?: McpCurrentOperationHint
}): string => {
  if (displayState === "checking" && currentOperation) {
    return `${getOperationLabel(currentOperation)} is in progress.`
  }

  switch (primaryReasonCode) {
    case undefined:
    case "partial_capability":
      return `${readyServerCount} of ${totalServers} MCP server${
        totalServers === 1 ? "" : "s"
      } ready.`
    case "not_configured":
      return "Add an MCP server to begin setup."
    case "preflight_failed":
      return "One or more MCP servers failed preflight validation."
    case "auth_missing":
      return "One or more MCP servers are missing required credentials."
    case "runtime_unavailable":
      return "One or more MCP server runtimes are unavailable."
    case "unreachable":
      return "One or more MCP servers are unreachable."
    case "discovery_failed":
      return "One or more MCP servers failed tool discovery."
    case "config_changed":
      return "One or more MCP server configurations changed after discovery."
    case "catalog_expired":
      return "One or more MCP tool catalogs are expired."
    case "discovery_not_run":
      return "Run discovery to populate MCP tool catalogs."
    case "no_tools_returned":
      return "Discovery completed, but one or more MCP servers returned no tools."
  }
}

const countMatchingTools = (
  server: McpHubExternalServer,
  registryEntries: McpHubToolRegistryEntry[]
): number => {
  const externalModule = `external.${server.id}`
  const externalToolPrefix = `ext.${server.id}.`

  return registryEntries.filter((entry) => {
    const hasExternalToolName = entry.tool_name.startsWith(externalToolPrefix)

    // Built-in modules can share ids with external servers; require an external
    // namespace signal before counting a registry row for server readiness.
    return hasExternalToolName || entry.module === externalModule
  }).length
}

const getReadinessHint = (
  readinessHintsByServerId: ReadinessHintsByServerId | undefined,
  serverId: string
): McpServerReadinessHint | undefined => {
  if (!readinessHintsByServerId) {
    return undefined
  }

  if (readinessHintsByServerId instanceof Map) {
    return readinessHintsByServerId.get(serverId)
  }

  return readinessHintsByServerId[serverId]
}

const getAuthTemplateBlockedReason = (
  server: McpHubExternalServer
): string | undefined => {
  const blockedReason = server.auth_template_blocked_reason?.trim()

  return blockedReason || undefined
}

const allowsNoAuthTemplate = (server: McpHubExternalServer): boolean => {
  const blockedReason = getAuthTemplateBlockedReason(server)

  return !blockedReason || blockedReason === "no_auth_template"
}

const hasBlockingAuthTemplateIssue = (server: McpHubExternalServer): boolean => {
  return (
    (server.auth_template_present === true && server.auth_template_valid === false) ||
    !allowsNoAuthTemplate(server)
  )
}

const isOperationalManagedServer = (server: McpHubExternalServer): boolean =>
  server.enabled !== false &&
  server.server_source !== "legacy" &&
  !server.superseded_by_server_id

export const getMcpCredentialState = (
  server: McpHubExternalServer
): McpCredentialState => {
  const credentialSlots = server.credential_slots ?? []
  const hasAuthTemplate = server.auth_template_present === true

  if (credentialSlots.some((slot) => slot.is_required && !slot.secret_configured)) {
    return "required_missing"
  }

  if (
    credentialSlots.some((slot) => slot.secret_configured) ||
    server.auth_template_valid === true
  ) {
    return "configured"
  }

  if (server.secret_configured && !hasAuthTemplate && credentialSlots.length === 0) {
    return "legacy_fallback"
  }

  if (
    server.transport === "stdio" &&
    !hasAuthTemplate &&
    credentialSlots.length === 0 &&
    allowsNoAuthTemplate(server)
  ) {
    return "not_required"
  }

  return "unknown"
}

export const getMcpServerReadiness = ({
  server,
  registryEntries,
  readinessHint
}: {
  server: McpHubExternalServer
  registryEntries: McpHubToolRegistryEntry[]
  readinessHint?: McpServerReadinessHint
}): McpServerReadiness => {
  const credentialState = getMcpCredentialState(server)
  const toolCount = countMatchingTools(server, registryEntries)
  const unsortedReasonCodes: McpReasonCode[] = []

  if (readinessHint?.preflightFailed || hasBlockingAuthTemplateIssue(server)) {
    unsortedReasonCodes.push("preflight_failed")
  }

  if (credentialState === "required_missing") {
    unsortedReasonCodes.push("auth_missing")
  }

  if (server.enabled === false || server.runtime_executable === false) {
    unsortedReasonCodes.push("runtime_unavailable")
  }

  if (readinessHint?.unreachable) {
    unsortedReasonCodes.push("unreachable")
  }

  if (readinessHint?.discoveryFailed) {
    unsortedReasonCodes.push("discovery_failed")
  }

  if (readinessHint?.configChanged) {
    unsortedReasonCodes.push("config_changed")
  }

  if (readinessHint?.catalogExpired) {
    unsortedReasonCodes.push("catalog_expired")
  }

  if (toolCount === 0) {
    unsortedReasonCodes.push(
      readinessHint?.discoverySucceededWithNoTools
        ? "no_tools_returned"
        : "discovery_not_run"
    )
  }

  if (readinessHint?.partialCapability && toolCount > 0) {
    unsortedReasonCodes.push("partial_capability")
  }

  const reasonCodes = sortReasonCodes(unsortedReasonCodes)
  const primaryReasonCode = reasonCodes[0]
  const currentOperation = readinessHint?.currentOperation ?? undefined
  const displayState = currentOperation
    ? "checking"
    : getDisplayStateForReason(primaryReasonCode)
  const allowedActions = currentOperation
    ? CHECKING_ACTIONS
    : getActionsForReasons(reasonCodes, SERVER_READY_ACTIONS)

  return {
    serverId: server.id,
    serverName: server.name,
    displayState,
    credentialState,
    toolCount,
    reasonCodes,
    primaryReasonCode,
    allowedActions,
    message: getMessageForServer({
      displayState,
      primaryReasonCode,
      credentialState,
      toolCount,
      currentOperation
    }),
    ...(currentOperation ? { currentOperation } : {})
  }
}

export const getMcpHubReadiness = ({
  servers,
  registryEntries,
  readinessHintsByServerId
}: {
  servers: McpHubExternalServer[]
  registryEntries: McpHubToolRegistryEntry[]
  readinessHintsByServerId?: ReadinessHintsByServerId
}): McpHubReadiness => {
  const operationalServers = servers.filter(isOperationalManagedServer)

  if (operationalServers.length === 0) {
    const reasonCodes = sortReasonCodes(["not_configured"])
    const primaryReasonCode = reasonCodes[0]

    return {
      displayState: "needs_setup",
      reasonCodes,
      primaryReasonCode,
      allowedActions: getActionsForReasons(reasonCodes),
      message: getMessageForHub({
        displayState: "needs_setup",
        primaryReasonCode,
        totalServers: 0,
        readyServerCount: 0
      }),
      servers: [],
      totalServers: 0,
      readyServerCount: 0,
      checkingServerCount: 0,
      attentionServerCount: 0,
      noToolServerCount: 0,
      staleServerCount: 0
    }
  }

  const serverReadiness = operationalServers.map((server) =>
    getMcpServerReadiness({
      server,
      registryEntries,
      readinessHint: getReadinessHint(readinessHintsByServerId, server.id)
    })
  )
  const reasonCodes = sortReasonCodes(
    serverReadiness.flatMap((readiness) => readiness.reasonCodes)
  )
  const primaryReasonCode = reasonCodes[0]
  const currentOperation = serverReadiness.find(
    (readiness) => readiness.currentOperation
  )?.currentOperation
  const displayState = currentOperation
    ? "checking"
    : getDisplayStateForReason(primaryReasonCode)
  const readyServerCount = serverReadiness.filter(
    (readiness) => readiness.displayState === "ready"
  ).length

  return {
    displayState,
    reasonCodes,
    primaryReasonCode,
    allowedActions: currentOperation
      ? CHECKING_ACTIONS
      : getActionsForReasons(reasonCodes, HUB_READY_ACTIONS),
    message: getMessageForHub({
      displayState,
      primaryReasonCode,
      totalServers: operationalServers.length,
      readyServerCount,
      currentOperation
    }),
    servers: serverReadiness,
    totalServers: operationalServers.length,
    readyServerCount,
    checkingServerCount: serverReadiness.filter(
      (readiness) => readiness.displayState === "checking"
    ).length,
    attentionServerCount: serverReadiness.filter(
      (readiness) => readiness.displayState === "needs_attention"
    ).length,
    noToolServerCount: serverReadiness.filter(
      (readiness) => readiness.displayState === "no_tools"
    ).length,
    staleServerCount: serverReadiness.filter(
      (readiness) => readiness.displayState === "stale"
    ).length
  }
}

File: tldw_Server_API/app/core/MCP_unified/mcp_hub_readiness.py
"""MCP Hub readiness policy helpers with sanitized response payload builders."""

from __future__ import annotations

from typing import Any

_READINESS_REASON_PRIORITY = [
    "auth_missing",
    "runtime_unavailable",
    "preflight_failed",
    "unreachable",
    "discovery_failed",
    "config_changed",
    "discovery_not_run",
    "no_tools_returned",
    "catalog_expired",
    "partial_capability",
]
_DISPLAY_STATE_BY_REASON = {
    "not_configured": "needs_setup",
    "preflight_failed": "needs_attention",
    "discovery_not_run": "needs_attention",
    "auth_missing": "needs_attention",
    "runtime_unavailable": "needs_attention",
    "unreachable": "needs_attention",
    "discovery_failed": "needs_attention",
    "no_tools_returned": "no_tools",
    "config_changed": "stale",
    "catalog_expired": "stale",
    "partial_capability": "ready",
}
_ACTIONS_BY_REASON = {
    "not_configured": ["add_server"],
    "preflight_failed": ["edit_config", "validate", "view_details"],
    "discovery_not_run": ["refresh_discovery", "edit_config"],
    "auth_missing": ["open_credentials", "view_details"],
    "runtime_unavailable": ["edit_config", "view_details"],
    "unreachable": ["edit_config", "refresh_discovery", "view_details"],
    "discovery_failed": ["refresh_discovery", "view_details"],
    "no_tools_returned": ["refresh_discovery", "view_details"],
    "config_changed": ["refresh_discovery", "edit_config"],
    "catalog_expired": ["refresh_discovery", "view_details"],
    "partial_capability": ["open_tool_catalog", "view_details"],
}


def _readiness_unique_reasons(reasons: list[str]) -> list[str]:
    """Return reason codes ordered by MCP Hub readiness priority."""
    seen = set()
    ordered: list[str] = []
    priority = {reason: index for index, reason in enumerate(_READINESS_REASON_PRIORITY)}
    for reason in sorted(reasons, key=lambda value: priority.get(value, len(priority))):
        if reason not in seen:
            seen.add(reason)
            ordered.append(reason)
    return ordered


def _readiness_primary_reason(reasons: list[str]) -> str | None:
    """Return the highest-priority readiness reason, if any."""
    ordered = _readiness_unique_reasons(reasons)
    return ordered[0] if ordered else None


def _readiness_allowed_actions(reasons: list[str]) -> list[str]:
    """Union readiness actions for the supplied reason codes."""
    actions: list[str] = []
    for reason in reasons:
        for action in _ACTIONS_BY_REASON.get(reason, []):
            if action not in actions:
                actions.append(action)
    return actions


def _readiness_message(primary_reason: str | None, credential_state: str) -> str:
    """Return a user-facing readiness status message without exposing config details."""
    if primary_reason == "auth_missing":
        return "Credentials are required before this server can be used."
    if primary_reason == "runtime_unavailable":
        return "Runtime is not available for this server."
    if primary_reason == "preflight_failed":
        return "Preflight validation failed. Check the server configuration."
    if primary_reason == "unreachable":
        return "Server cannot be reached."
    if primary_reason == "discovery_failed":
        return "Discovery ran but failed."
    if primary_reason == "config_changed":
        return "Server config or discovery state changed. Refresh discovery."
    if primary_reason == "discovery_not_run":
        if credential_state == "not_required":
            return "No credentials required. Discover tools to make this server available."
        return "Server is saved, but tool discovery has not run."
    if primary_reason == "no_tools_returned":
        return "Server responded, but exposed no tools."
    if primary_reason == "catalog_expired":
        return "Tool catalog is stale. Refresh discovery."
    if primary_reason == "partial_capability":
        return "Ready with limited capability."
    if primary_reason == "not_configured":
        return "Add an external server to start MCP Hub setup."
    return "Ready. No credentials required." if credential_state == "not_required" else "Ready."


def _credential_state_for_external_row(row: dict[str, Any]) -> str:
    """Normalize external-server credential state without returning secret material."""
    slots = [slot for slot in (row.get("credential_slots") or []) if isinstance(slot, dict)]
    required_slots = [slot for slot in slots if bool(slot.get("is_required", True))]
    if any(not bool(slot.get("secret_configured")) for slot in required_slots):
        return "required_missing"
    if any(bool(slot.get("secret_configured")) for slot in slots) or row.get("auth_template_valid") is True:
        return "configured"
    has_auth_template = bool(row.get("auth_template_present"))
    if bool(row.get("secret_configured")) and not has_auth_template and not slots:
        return "legacy_fallback"
    blocked_reason = str(row.get("auth_template_blocked_reason") or "").strip()
    if (
        str(row.get("transport") or "").strip().lower() == "stdio"
        and not has_auth_template
        and not slots
        and blocked_reason in {"", "no_auth_template"}
    ):
        return "not_required"
    return "unknown"


def _is_authoritative_external_tool(entry: dict[str, Any], server_id: str) -> bool:
    """Return True only for registry signals that authoritatively belong to an external server."""
    tool_name = str(entry.get("tool_name") or "")
    module_name = str(entry.get("module") or "")
    return tool_name.startswith(f"ext.{server_id}.") or module_name == f"external.{server_id}"


def _matching_external_tool_entries(
    registry_entries: list[dict[str, Any]],
    *,
    server_id: str,
) -> list[dict[str, Any]]:
    """Return registry entries that belong to one external server."""
    return [entry for entry in registry_entries if _is_authoritative_external_tool(entry, server_id)]


def _operation_metadata_from_row(row: dict[str, Any]) -> dict[str, Any] | None:
    """Return in-flight operation metadata when the row exposes it."""
    current_operation = row.get("current_operation")
    return current_operation if isinstance(current_operation, dict) else None


def _coerce_nonnegative_int(value: Any) -> int:
    """Coerce untrusted numeric refresh metadata to a non-negative integer."""
    try:
        return max(0, int(value))
    except (TypeError, ValueError):
        return 0


def _extract_refresh_payload(raw_result: dict[str, Any]) -> dict[str, Any]:
    """Extract the manager refresh payload from direct or MCP content-wrapped results."""
    if not isinstance(raw_result, dict):
        return {}
    if any(key in raw_result for key in ("refreshed_servers", "total_servers", "virtual_tools", "errors")):
        return raw_result
    result = raw_result.get("result")
    if isinstance(result, dict):
        extracted = _extract_refresh_payload(result)
        if extracted:
            return extracted
    content = raw_result.get("content")
    if isinstance(content, list):
        for item in content:
            if not isinstance(item, dict):
                continue
            json_payload = item.get("json")
            if isinstance(json_payload, dict):
                return json_payload
    return {}


def sanitize_discovery_refresh_result_payload(raw_result: dict[str, Any]) -> dict[str, Any]:
    """Return refresh metadata with generic error messages only."""
    payload = _extract_refresh_payload(raw_result)
    raw_errors = payload.get("errors") if isinstance(payload.get("errors"), dict) else {}
    errors = {str(server_id): "Discovery refresh failed." for server_id in raw_errors}
    return {
        "refreshed_servers": _coerce_nonnegative_int(payload.get("refreshed_servers")),
        "total_servers": _coerce_nonnegative_int(payload.get("total_servers")),
        "virtual_tools": _coerce_nonnegative_int(payload.get("virtual_tools")),
        "errors": errors,
    }


def build_server_readiness_payload(
    row: dict[str, Any],
    *,
    registry_entries: list[dict[str, Any]],
    current_operation: dict[str, Any] | None = None,
    last_validation_at: Any | None = None,
    refresh_result: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Build sanitized readiness payload for one external server row."""
    server_id = str(row.get("id") or "")
    server_name = str(row.get("name") or server_id)
    credential_state = _credential_state_for_external_row(row)
    matching_entries = _matching_external_tool_entries(registry_entries, server_id=server_id)
    tool_count = len(matching_entries)
    operation = current_operation or _operation_metadata_from_row(row)
    if operation:
        operation_type = str(operation.get("operation_type") or "validation")
        if operation_type not in {"validation", "discovery"}:
            operation_type = "validation"
        return {
            "server_id": server_id,
            "server_name": server_name,
            "display_state": "checking",
            "credential_state": credential_state,
            "tool_count": tool_count,
            "reason_codes": [],
            "primary_reason_code": None,
            "allowed_actions": ["view_details"],
            "message": (
                "Tool discovery is running."
                if operation_type == "discovery"
                else "Preflight validation is running."
            ),
            "current_operation": {
                "operation_type": operation_type,
                "started_at": operation.get("started_at"),
                "message": operation.get("message"),
            },
            "last_validation_at": last_validation_at or row.get("last_validation_at"),
            "last_discovery_at": row.get("last_discovery_at"),
            "last_successful_discovery_at": row.get("last_successful_discovery_at"),
        }

    reasons: list[str] = []
    if credential_state == "required_missing":
        reasons.append("auth_missing")
    if not bool(row.get("enabled")) or row.get("runtime_executable") is False:
        reasons.append("runtime_unavailable")

    blocked_reason = row.get("auth_template_blocked_reason")
    if (row.get("auth_template_present") and row.get("auth_template_valid") is False) or (
        blocked_reason and str(blocked_reason) != "no_auth_template"
    ):
        reasons.append("preflight_failed")

    last_error_category = row.get("last_error_category")
    last_error_message = row.get("last_error_message")
    refresh_errors = refresh_result.get("errors") if isinstance(refresh_result, dict) else {}
    if refresh_result is not None:
        if isinstance(refresh_errors, dict) and server_id in refresh_errors:
            reasons.append("discovery_failed")
            last_error_category = "discovery_failed"
            last_error_message = "Discovery refresh failed."
        elif last_error_category == "discovery_failed":
            last_error_category = None
            last_error_message = None
    elif last_error_category == "discovery_failed":
        reasons.append("discovery_failed")
        last_error_message = "Discovery refresh failed."

    if tool_count == 0 and "runtime_unavailable" not in reasons and "discovery_failed" not in reasons:
        reasons.append("discovery_not_run")
    if tool_count > 0 and any(entry.get("metadata_warnings") for entry in matching_entries):
        reasons.append("partial_capability")

    reason_codes = _readiness_unique_reasons(reasons)
    primary_reason = _readiness_primary_reason(reason_codes)
    display_state = _DISPLAY_STATE_BY_REASON.get(primary_reason or "", "ready")
    allowed_actions = (
        _readiness_allowed_actions(reason_codes)
        if reason_codes
        else ["open_tool_catalog", "view_details"]
    )
    return {
        "server_id": server_id,
        "server_name": server_name,
        "display_state": display_state,
        "credential_state": credential_state,
        "tool_count": tool_count,
        "reason_codes": reason_codes,
        "primary_reason_code": primary_reason,
        "allowed_actions": allowed_actions,
        "message": _readiness_message(primary_reason, credential_state),
        "current_operation": None,
        "last_validation_at": last_validation_at or row.get("last_validation_at"),
        "last_discovery_at": row.get("last_discovery_at"),
        "last_successful_discovery_at": row.get("last_successful_discovery_at"),
        "last_error_category": last_error_category,
        "last_error_message": last_error_message,
        "refresh_result": refresh_result,
    }


def is_operational_managed_external_row(row: dict[str, Any]) -> bool:
    """Return True when a row counts toward active MCP Hub readiness."""
    return (
        str(row.get("server_source") or "managed") == "managed"
        and not row.get("superseded_by_server_id")
        and bool(row.get("enabled"))
    )


def is_available_for_discovery_refresh(row: dict[str, Any]) -> bool:
    """Return True when the server can run discovery refresh from the current process."""
    return is_operational_managed_external_row(row) and row.get("runtime_executable") is True


def build_hub_readiness_payload(
    rows: list[dict[str, Any]],
    registry_entries: list[dict[str, Any]],
) -> dict[str, Any]:
    """Build sanitized aggregate MCP Hub readiness payload for visible external servers."""
    server_readiness = [
        build_server_readiness_payload(row, registry_entries=registry_entries)
        for row in rows
    ]
    operational_server_ids = {
        str(row.get("id") or "")
        for row in rows
        if is_operational_managed_external_row(row)
    }
    operational_readiness = [
        readiness
        for readiness in server_readiness
        if readiness["server_id"] in operational_server_ids
    ]

    if not operational_readiness:
        return {
            "display_state": "needs_setup",
            "reason_codes": ["not_configured"],
            "primary_reason_code": "not_configured",
            "allowed_actions": ["add_server"],
            "message": _readiness_message("not_configured", "not_required"),
            "servers": server_readiness,
            "total_servers": 0,
        }

    aggregate_reasons = _readiness_unique_reasons(
        [
            reason
            for readiness in operational_readiness
            for reason in readiness["reason_codes"]
        ]
    )
    primary_reason = _readiness_primary_reason(aggregate_reasons)
    if any(readiness["display_state"] == "checking" for readiness in operational_readiness):
        display_state = "checking"
    elif primary_reason:
        display_state = _DISPLAY_STATE_BY_REASON.get(primary_reason, "needs_attention")
    else:
        display_state = "ready"

    return {
        "display_state": display_state,
        "reason_codes": aggregate_reasons,
        "primary_reason_code": primary_reason,
        "allowed_actions": (
            _readiness_allowed_actions(aggregate_reasons)
            if aggregate_reasons
            else ["open_tool_catalog", "view_details"]
        ),
        "message": (
            _readiness_message(primary_reason, "configured")
            if primary_reason
            else "MCP Hub is ready."
        ),
        "servers": server_readiness,
        "total_servers": len(operational_readiness),
        "ready_server_count": sum(1 for item in operational_readiness if item["display_state"] == "ready"),
        "checking_server_count": sum(1 for item in operational_readiness if item["display_state"] == "checking"),
        "attention_server_count": sum(
            1 for item in operational_readiness if item["display_state"] == "needs_attention"
        ),
        "no_tool_server_count": sum(1 for item in operational_readiness if item["display_state"] == "no_tools"),
        "stale_server_count": sum(1 for item in operational_readiness if item["display_state"] == "stale"),
    }
